src/adapter/src/rbac.rs

```diff
@@ -494,73 +510,77 @@ fn generate_required_privileges(
                 .into_iter()
             })
             .collect(),
-        Plan::CreateSecret(plan) => BTreeSet::from([(
+        Plan::CreateSecret(plan) => vec![(
             plan.name.qualifiers.clone().into(),
             AclMode::CREATE,
             role_id,
-        )]),
+        )],
         Plan::CreateSink(plan) => {
-            let mut privileges = BTreeSet::from([(
-                plan.name.qualifiers.clone().into(),
-                AclMode::CREATE,
+            let schema_id: ObjectId = plan.name.qualifiers.clone().into();
+            let mut privileges = vec![(schema_id.clone(), AclMode::CREATE, role_id)];
+            privileges.extend_from_slice(&generate_read_privileges_inner(
+                catalog,
+                iter::once(plan.sink.from),
                 role_id,
-            )]);
+                &mut BTreeSet::from([(schema_id, role_id)]),
+            ));
             if let Some(id) = plan.cluster_config.cluster_id() {
-                privileges.insert((id.into(), AclMode::CREATE, role_id));
+                privileges.push((id.into(), AclMode::CREATE, role_id));
```

Comment:
Uncovered, seems interesting.

Reply:
I have not ignored this comment, I'm trying to figure out how to test sinks without using testdrive. Though, I might just switch to testdrive.

Reply:
Just pushed some tests for this (as well as fixed some bugs I found while writing the test).